### PR TITLE
[netcore] Loader cleanup and LoadInDefaultContext test fixes

### PR DESCRIPTION
**Pull request by CoffeeFlux**
Should probably rebase this instead of squashing - the changes are largely isolated but I didn't want to make two pull requests.

The latter commit fixes the main remaining issue with the System.Runtime.Loader.Tests.DefaultLoadContextTests.LoadInDefaultContext test. Sadly, the test still can't be re-enabled because it fails the very first assert, which checks that a target assembly has never been loaded before, but should be otherwise functional.

Hopefully the order shuffling on the search hooks will not cause issues with legacy Mono, but we'll see.

### CODE REVIEW

**Comment by CoffeeFlux** [comment]
In case it's of any use, this is the Program.cs I used to run the test outside of the test harness itself: https://gist.github.com/CoffeeFlux/bd2f96c4975c49304765872b6a067d55

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -2467,9 +2467,14 @@ ves_icall_System_Reflection_Assembly_InternalLoad (MonoStringHandle name_handle,
...
 	char *name;
 
+	MonoAssembly *requesting_assembly = mono_runtime_get_caller_from_stack_mark (stack_mark);
 	MonoAssemblyLoadContext *alc = (MonoAssemblyLoadContext *)load_Context;
+
+	if (!alc)
+		alc = mono_assembly_get_alc (requesting_assembly);
 	if (!alc)
-		alc = mono_domain_default_alc (mono_domain_get ());
+		g_assert_not_reached ();
```
Comment (lambdageek) [inline mono/metadata/appdomain.c]: this is `g_assert (alc)`.  The same as line 2479.  Maybe just move that up to 2475?

Comment (CoffeeFlux) [reply mono/metadata/appdomain.c]: Ah whoops, yep.

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Comment by CoffeeFlux** [comment]
@monojenkins build failed

**Pull request merged**